components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/ApplicationsApiServiceImpl.java

```diff
@@ -1280,11 +1284,43 @@ public Response applicationsApplicationIdOauthKeysKeyMappingIdPut(String applica
             throws APIManagementException {
 
         String username = RestApiCommonUtil.getLoggedInUsername();
-            APIConsumer apiConsumer = APIManagerFactory.getInstance().getAPIConsumer(username);
-            Application application = apiConsumer.getApplicationByUUID(applicationId);
-            if (application != null) {
-                ApplicationKeyDTO appKey = getApplicationKeyByAppIDAndKeyMapping(applicationId, keyMappingId);
-                if (RestAPIStoreUtils.isUserOwnerOfApplication(application) && appKey != null) {
+        APIConsumer apiConsumer = APIManagerFactory.getInstance().getAPIConsumer(username);
+        String organization = RestApiUtil.getValidatedOrganization(messageContext);
+        Application application = apiConsumer.getApplicationByUUID(applicationId);
+        if (application != null) {
+            ApplicationKeyDTO appKey = getApplicationKeyByAppIDAndKeyMapping(applicationId, keyMappingId);
+            if (RestAPIStoreUtils.isUserOwnerOfApplication(application) && appKey != null) {
+                // Mapped Keys are from external IDPs
+                OAuthApplicationInfo updatedData = null;
+                if (ApplicationKeyDTO.ModeEnum.MAPPED.equals(appKey.getMode())) {
+                    updatedData = apiConsumer.updateMappedApplicationKey(username, application,
+                            appKey.getKeyType().value(), appKey.getKeyManager(), body.getConsumerKey(), organization);
+
+                    ApplicationRegistrationEvent applicationKeyRemoveEvent = new ApplicationRegistrationEvent(
+                            UUID.randomUUID().toString(), System.currentTimeMillis(),
+                            APIConstants.EventType.REMOVE_APPLICATION_KEYMAPPING.name(),
+                            MultitenantConstants.SUPER_TENANT_ID, organization, application.getId(),
+                            application.getUUID(), appKey.getConsumerKey(),
+                            appKey.getKeyType().value().toUpperCase(), appKey.getKeyManager());
+                    APIUtil.sendNotification(applicationKeyRemoveEvent,
+                            APIConstants.NotifierType.APPLICATION_REGISTRATION.name());
+                    ApplicationRegistrationEvent applicationRegistrationEvent = new ApplicationRegistrationEvent(
+                            UUID.randomUUID().toString(), System.currentTimeMillis(),
+                            APIConstants.EventType.APPLICATION_REGISTRATION_CREATE.name(),
+                            MultitenantConstants.SUPER_TENANT_ID, organization, application.getId(),
+                            application.getUUID(), updatedData.getClientId(),
+                            updatedData.getTokenType(), appKey.getKeyManager());
+                    APIUtil.sendNotification(applicationRegistrationEvent,
+                            APIConstants.NotifierType.APPLICATION_REGISTRATION.name());
```

Comment:
shall we move the notification related logic to the updateMappedApplicationKey method?